### ISSUE TITLE
Native RDOS linker

### ISSUE THREAD

**Issue by rdos314** [open-watcom/open-watcom-v2#415]
This test was made with a simple "hello world" program:
It's compiled like this: "wcc386 hello"

Environment variables are set like this:
SET INCLUDE=d:\h
SET LIB=d:\lib

The include files from OW rel/h are copied to the target machine and selective libraries from OW rel/lib386 are also copied.

When used with system specification, everything seems fine regardless if lib path is used or not:
wlink name hello file hello sys rdos
Open Watcom Linker Version 2.0 beta Mar 24 2018 23:15:17 (32-bit)
Copyright (c) 2002-2018 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
Warning! W1115: system block rdos: environment name WATCOM not found
Warning! W1115: system block rdos: environment name WATCOM not found
loading object files
searching libraries
creating a RDOS executable

When used without system specification and MK_PE as a hint:
wlink name hello file hello
Open Watcom Linker Version 2.0 beta Mar 24 2018 23:15:17 (32-bit)
Copyright (c) 2002-2018 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
loading object files
searching libraries
Warning! W1008: cannot open emu387.lib : No error
Warning! W1008: cannot open math387r.lib : No error
Warning! W1008: cannot open clib3r.lib : No error
Error! E2028: __CHK is an undefined reference
Error! E2028: printf_ is an undefined reference
Error! E2028: _cstart_ is an undefined reference
Warning! W1023: no starting address found, using 00401000
creating a PE executable
file hello.obj(D:\bin\hello.c): undefined symbol __CHK
file hello.obj(D:\bin\hello.c): undefined symbol printf_

When used without system specification, MK_PE as a hint, but forcing the linker to always use the LIBPATH (the executable works but it is a PE file with some inconsistency):
Open Watcom Linker Version 2.0 beta Mar 24 2018 23:15:17 (32-bit)
Copyright (c) 2002-2018 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
loading object files
searching libraries
creating a PE executable
Warning! W1060: value for alignment incorrect

When used without system specification, MK_RDOS as a hint, but forcing the linker to always use the LIBPATH (the executable doesn't work and it seems to be in some 32-bit segmented format):
Open Watcom Linker Version 2.0 beta Mar 24 2018 23:15:17 (32-bit)
Copyright (c) 2002-2018 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
loading object files
searching libraries
creating a RDOS executable

When used without system specification, MK_RDOS as a hint.
Open Watcom Linker Version 2.0 beta Mar 24 2018 23:15:17 (32-bit)
Copyright (c) 2002-2018 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
loading object files
searching libraries
Warning! W1008: cannot open emu387.lib : No error
Warning! W1008: cannot open math387r.lib : No error
Warning! W1008: cannot open clib3r.lib : No error
Error! E2028: __CHK is an undefined reference
Error! E2028: printf_ is an undefined reference
Error! E2028: _cstart_ is an undefined reference
Warning! W1023: no starting address found, using 0001:00000000
creating a RDOS executable
file hello.obj(D:\bin\hello.c): undefined symbol __CHK
file hello.obj(D:\bin\hello.c): undefined symbol printf_

**Comment by jmalak** [comment]
I think the problem is caused by missing RDOS specific "default" targets "286" and "386" in wlink.sp (source for wlink.lnk file). 
Try to add this two targets "286" for default 16-bit target and "386" for default 32-bit target to wlink.sp source file and rebuild wlink.lnk and use it for testing. As guide you can look on other targets definition for "286" and "386" targets.

**Comment by rdos314** [comment]
I reverted the previous changes + added RDOS as default target. It seems a bit better, but it still doesn't work with default target.

This is how it looks like without the MK_RDOS suggestion:

Open Watcom Linker Version 2.0 beta Mar 26 2018 09:48:01 (32-bit)
Copyright (c) 2002-2018 The Open Watcom Contributors. All Rights Reserved.
Portions Copyright (c) 1985-2002 Sybase, Inc. All Rights Reserved.
Source code is available under the Sybase Open Watcom Public License.
See http://www.openwatcom.org/ for details.
loading object files
Warning! W1058: system block hello.obj: windows option not valid for a rdos executable
Warning! W1111: system block hello.obj: error in default system block
searching libraries
Error! E3002: ** internal ** - format not decided

The format in the default system is "windows pe rdos", exactly the same as in the system rdos section. Also tried it with "rdos" (same error), and with "pe" (gives a more fatal error).

When adding the MK_RDOS suggestion, the results are like before with libraries not found.

**Comment by jmalak** [comment]
Warning! W1058: system block hello.obj: windows option not valid for a rdos executable
It looks like wlink decide for correct format but wlink.lnk is damaged or incorrect version..

Try to debug how decision about executable format is now going if target is not specified.

May be that some asumption or rule is wrong for RDOS.
Take into account if MK_RDOS is used then you must probably correct something in formats.h for RDOS.
You must check it for 32-bit and also for 16-bit.

**Comment by jmalak** [comment]
Check wlink full command, may be some wrong options are there.

**Comment by jmalak** [comment]
Now you have defined RDOS format for RDOS, but you use WINDOWS format in wlink scripts.
You should use RDOS as format, not windows.

**Comment by jmalak** [comment]
It looks like a problem with RDOS executable format options.
In documentation is specified that it should uses format "WINDOWS RDOS", but in wlink it is defined as format "RDOS" or "WINDOWS RDOS", but format "WINDOWS RDOS" is not enabled for RDOS.
It will require to decide what form will be used for RDOS and fix wlink and wlink scripts to use this form.

**Comment by rdos314** [comment]
But the format is a subset of windows PE format with a different header.

OK, so if I change wlink script to use "rdos", it actually seems to work. It runs just fine without any errors, but the executable doesn't work. I think that is because RDOS has no specific format. 

OTOH, I can add MK_RDOS to Windows format, and then it accepts the windows format, but then complains about "pe" instead.

**Comment by jmalak** [comment]
I studied a little bit wlink code for RDOS.
There is one confusion which I overlooked.
`format RDOS` is for rdos specific binary formats (drivers etc.), but not used by regular RDOS executable.
It is undocumented in WLINK guide. It should be updated.
`MK_RDOS macro` default to this binary formats.
The regular RDOS executable is `format WINDOWS RDOS`, it is handled by `MK_PE macro`.
Therefore if it should default to regular RDOS format MK_PE macro should be used for default and scripts should refere to format WINDOWS RDOS.
Use of `format RDOS` must be always explicit, but implicit should be `format WINDOWS RDOS`.

I think LAST_CHANCE macro for RDOS should be (MK_PE | MK_DOS_EXE) as minimum.
